lib/appsignal/rack/body_wrapper.rb

```diff
@@ -0,0 +1,128 @@
+# frozen_string_literal: true
+
+module Appsignal
+  # @api private
+  module Rack
+    class BodyWrapper
+      def self.wrap(original_body, appsignal_transaction_or_nil)
+        # The logic of how Rack treats a response body differs based on which methods
+        # the body responds to. This means that to support the Rack 3.x spec in full
+        # we need to return a wrapper which matches the API of the wrapped body as closely
+        # as possible. Pick the wrapper from the most specific to the least specific.
+        # See https://github.com/rack/rack/blob/main/SPEC.rdoc#the-body-
+        #
+        # What is important is that our Body wrapper responds to the same methods Rack
+        # (or a webserver) would be checking and calling, and passes through that functionality
+        # to the original body. This can be done using delegation via i.e. SimpleDelegate
+        # but we also need "close" to get called correctly so that the Appsignal transaction
+        # gets completed - which will not happen, for example, when #to_ary gets called
+        # just on the delegated Rack body.
+        #
+        # This comment https://github.com/rails/rails/pull/49627#issuecomment-1769802573
+        # is of particular interest to understand why this has to be somewhat complicated.
+        if original_body.respond_to?(:to_path)
+          PathableBodyWrapper.new(original_body, appsignal_transaction_or_nil)
+        elsif original_body.respond_to?(:to_ary)
+          ArrayableBodyWrapper.new(original_body, appsignal_transaction_or_nil)
+        elsif !original_body.respond_to?(:each) && original_body.respond_to?(:call)
+          CallableBodyWrapper.new(original_body, appsignal_transaction_or_nil)
+        else
+          EnumerableBodyWrapper.new(original_body, appsignal_transaction_or_nil)
+        end
+      end
+
+      def initialize(body, appsignal_transaction)
+        @body_already_closed = false
+        @body = body
+        @transaction = appsignal_transaction
+      end
+
+      # This must be present in all Rack bodies and will be called by the serving adapter
+      def close
+        # The @body_already_closed check is needed so that if `to_ary`
+        # of the body has already closed itself (as prescribed) we do not
+        # attempt to close it twice
+        @body.close if !@body_already_closed && @body.respond_to?(:close)
+        @body_already_closed = true
+      rescue Exception => error # rubocop:disable Lint/RescueException
+        @transaction&.set_error(error)
+        raise error
+      ensure
+        @transaction&.complete
```

Comment:
You need to call `Appsignal::Transaction.complete_current!` so it unsets the current transaction stored in the Thread variable.

https://github.com/appsignal/appsignal-ruby/blob/4c2845cedf6399b973759e31a539a415d4bb565b/lib/appsignal/transaction.rb#L65-L74